### PR TITLE
[ramda] remove placeholder

### PR DESCRIPTION
**Pull request by adispring**
Ramda does not support TS well, one of the reason is TS does not support curry well enough, and `R.placeholder` make this worse, such as put a function with R.placeholder as a param into another function, in this case TS can not infer the type correct.

R.merge and R.contains have been removed since v0.28.0.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ramda ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ramda@latest/))
was missing the following properties:
1. collectBy
1. modify
1. modifyPath
1. partialObject
1. on

<details>
<summary>as well as these 5 other properties...</summary>
<p>promap, sequence, splitWhenever, unwind, whereAny</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against c18fd980cd9a503038ef215fd0cfdada5dd251ad
</p>

**Review by essenmitsosse — APPROVED** [review]
I can't really speak to the removal of `Placeholder` since I do not use it mayself for the reasons you stated und because I find the handling just weird anyway. Don't know what the implications are for other people.

So from my side, this is good to go.

**Review by essenmitsosse — APPROVED** [review]

**Comment by ycmjason** [comment]
> `R.placeholder` make this worse, such as put a function with R.placeholder as a param into another function, in this case TS can not infer the type correct.

Can you give an example how the overloads with placeholder will break the type inference?

**Comment by somebody1234** [comment]
(And has it been confirmed that moving placeholder overloads *below* normal overloads will not fix this?)

**Comment by ycmjason** [comment]
((will removing the placeholder overloads be a breaking change for code that is currently using them?))

**Comment by adispring** [comment]
> > `R.placeholder` make this worse, such as put a function with R.placeholder as a param into another function, in this case TS can not infer the type correct.
> 
> Can you give an example how the overloads with placeholder will break the type inference?

@ycmjason Only few ramda apis' types have `Placeholder` overloads, and these Placeholder` overloads can not support full ability of ramda's `Placeholder`. 

Even worse, some @types/ramda api's `Placeholder` overload implementation is wrong.

Ramda and Typescript is just tools help people do their jobs more efficiently and safely, not the opposite.

For example:

`R.someApi(R.__)(R.__)(R.__)(...)` is valid, we can write any number of `R.__`s if the api has not enough arguments to execute. but we can't do that in Typescript with ramda.

Here are some more examples:

1. R.assoc

```ts
R.assoc(R.__); // Argument of type 'Placeholder' is not assignable to parameter of type 'string'.ts(2345)
R.assoc(R.__, 'value'); // Argument of type 'Placeholder' is not assignable to parameter of type 'string'.ts(2345)
```

2. `R.assocPath`, R.prop, R.propOr .etc,  has the same problem.

3. R.concat

```ts
R.concat(R.__)('abc'); // Expected 2 arguments, but got 1.ts(2554)
```

4. R.includes

In `R.includes`, the `placeholder` overload implementation is wrong:

```ts
export function includes(__: Placeholder): (list: readonly string[] | string, s: string) => boolean;
export function includes<T>(__: Placeholder): (list: readonly T[], target: T) => boolean;
```

should be:

```
export function includes(__: Placeholder): (s: string, list: readonly string[] | string) => boolean;
export function includes<T>(__: Placeholder): (target: T, list: readonly T[]) => boolean;
```

**Comment by adispring** [comment]
Ramda itself may also want to remove `Placeholder` in the future, here are some discussions:

https://github.com/ramda/ramda/pull/2705

https://github.com/ramda/ramda/pull/2629

**Comment by essenmitsosse** [comment]
I gave it some more thought and I agree with @adispring.

I think the philosophy here should be: If you can't get it right, better leave it out. People use TypeScript to get more confidence in their code, the worst you want is a false sense of security or a feature that randomly works in some cases and in some it doesn't. It's a much clearer statement for a user to just know `Placeholder` is not supported, then guessing or even having to dive into the types themselves to figure out when it is and when it isn't.

I'd rather have this library focus on the things it can reasonably support and do those things well. Also everybody is free to fork these types and add placeholder support and release a ramda-types-with-partial-placeholder-support version of this library.

Additionally I read on the `ramda` repo the other day that the future of `Placeholder` is unclear anyway. But maybe @CrossEye could chime in and give his opinion on the matter.

**Comment by somebody1234** [comment]
**TL;DR**: these aren't very convincing reasons to remove placeholder - so at the very least we should get the opinion of the community first.

sourcegraph results may be useful to see how widely used placeholder is:
[8.7k results](https://sourcegraph.com/search?q=context:global+lang:typescript+%27ramda%27+count:all&patternType=literal) for `ramda`, [274](https://sourcegraph.com/search?q=context:global+lang:typescript+%27ramda%27+AND+R.__&patternType=literal) for `ramda AND R.__`, which works out to a little over 3% of repositories that use placeholder 
in js, [1.2k](https://sourcegraph.com/search?q=context:global+lang:javascript+%27ramda%27+AND+R.__+count:all&patternType=literal) out of [24.4k](https://sourcegraph.com/search?q=context:global+lang:javascript+%27ramda%27+count:all&patternType=literal) = ~5%, if conversion from js is a goal.


anyway re: placeholder overloads - i don't think they're particularly difficult to add, so i strongly disagree with using it as a reason they should be removed.

vaguely related - the types do not fully support ramda's api even without placeholder, since we do not support `()` currying which a) is supported by ramda and b) is supported by [an old set of typings](https://github.com/typed-typings/npm-ramda) for specifying type parameters. so does that mean we should remove *all* typings, since none of them are currently typed correctly?

as for wrong overload implementations, this is *not unique to placeholder* either. so again, it should not be a reason to remove placeholder.

**Comment by adispring** [comment]
TypeScript is just a static tools, I don't think breaking change will introduce much destructive effect, it will effect noting in run time.

People can easily replace `R.__` with lambda express by ts compiler check tips, if he/she is using ramda with ts.

**Comment by adispring** [comment]
> as for wrong overload implementations, this is not unique to placeholder either. so again, it should not be a reason to remove placeholder.

I agree with this, this should not be a reason for removing placeholder.

**Comment by somebody1234** [comment]
that's a fair point, *however* it's worth noting that the complexity will be hidden from the user - it's only really an issue for types maintainers.
it *would* drastically increase the number of overloads though, which is definitely a bad thing.

worth noting `R.flip` is a viable option for those that use `R.__`, however not sure how well it plays with generic functions (= basically all of them). a lambda is similarly far from ideal because of this

**Comment by adispring** [comment]
> vaguely related - the types do not fully support ramda's api even without placeholder, since we do not support () currying which a) is supported by ramda 

I don't think this is a restrict of typescript. In most of the time, if people write something without argument like this: `R.someApi()`, it mostly indicate that people have forgotten to pass in argument. Is there any reason that people intend to write `R.someApi()` instead of `R.someApi`? I can't imagine any use case.

On the contrary, I think this is an advantage of typescript that remind(force) people maybe forget pass argument when call ramda api with argument.

**Comment by adispring** [comment]
> b) is supported by [an old set of typings](https://github.com/typed-typings/npm-ramda) for specifying type parameter

I don't even know there exist  [an old set of typings](https://github.com/typed-typings/npm-ramda) before now.

I have jump into  [this npm github repo](https://github.com/typed-typings/npm-ramda), it is not update five years ago. I doubt is there any body still use this npm.

I then go to npmjs.com. find out there is no npm called types/npm-ramda, but just a npm called @vtex/npm-ramda, maybe this is really types/npm-ramda, and I find there is only 4 weekly downloads.

**Comment by somebody1234** [comment]
right, but i was focusing more on them advertising that it can be used for specifying type parameters, which i think *is* useful (in some cases)

**Comment by somebody1234** [comment]
to be fair, the upcoming typescript feature [instantiaton expressions](https://github.com/microsoft/TypeScript/pull/47607) mostly fixes that so maybe we can ignore it

**Comment by rbuckton** [comment]
I can get on board with dropping contains/merge, but I'm not sure I agree with dropping `Placeholder`, at least not until there has been more time to consider the ramifications. @adispring, would it be possible to split this into two PRs (i.e., make this one contains/merge removal only, and a separate one for `Placeholder`)? Then we could discuss the merits of `Placeholder` separately from this discussion.

File: types/ramda/index.d.ts [hunk 1]
```diff
@@ -76,12 +75,6 @@ import {
 
 export * from './tools';
 
-/**
- * Placeholder. When used with functions like curry, or op, the second argument is applied to the second
- * position, and it returns a function waiting for its first argument.
- */
-export const __: Placeholder;
```
Comment (rbuckton) [inline types/ramda/index.d.ts:83]: I would recommend putting the removal of `Placeholder` into a separate PR since its removal is more controversial than contains/merge.

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/ramda/index.d.ts (hunk 1 above)
Comment (adispring) [reply types/ramda/index.d.ts:83]: OK, I'll split this pr into 2 separate PRs.

**Comment by CrossEye** [comment]
@essenmitsosse:

> Additionally I think I read on the ramda repo the other day that the future of Placeholder is unclear anyway — but not 100% sure about that and can't find it anymore. But maybe @CrossEye could chime in and give his opinion on the matter.

There are certainly no immediate plans for removing the placeholder.  There is a certain distaste for it among the core team, and I think many of us would be happy to see it go.  But there is no obvious replacement for this without a radical change to the library.  That may happen one day, but it's not in the plans for 1.0 or 2.0.

I could see us disallowing the placeholder being passed as the last argument in any one call.  It makes no logical sense, and probably contributes to the difficulties typing functions here.  It's allowed because there was no reason to add complexity to the code in order to disallow it; and if it's allowed, there is no reasonable alternative but to treat it as a no-op.  But if this is a significant factor in making it hard to type Ramda, we could certainly at least fix that.

I really don't know TS well, and am not versed in the details of the typings files.  But would disallowing that make typing Ramda's more complex signatures more palatable?  Ramda has always been a JS library, and we haven't really even tried making concessions in order to work better with TS.  We probably still wouldn't do so for our core API design, but this is a minor implementation detail.  It would technically be a backward incompatibility, but as we (finally) approach 1.0, we're going to have a few of those, so this might be a good time to talk about it.

We could also do something to the actual value, if it helps.  It's an object with a special named property on it; that's it.  If it would help for it to be of some constructed type, we could probably do so, so long as we can keep that property.

**Comment by essenmitsosse** [comment]
@CrossEye thanks for the input, that is definitely helpful.

Regarding disallowing certain use cases for Placeholders in the JavaScript implementation: It say it doesn't really matter for the Typings. If we got the input that certain use cases aren't really intended or necessary, its perfectly fine to not support them in TypeScript, while they still might be valid when actually running the code.

Example:
```javascript
const add = (a, b) => a + b
```

Assuming this function is supposed to add numbers, its implementation would allow it to also concatenate strings. So when writing a JavaScript library you might want to add a run time check to avoid someone foot-gunning themself. But in TypeScript, you could type this function as only being allowed to take a number — making it actually easier to narrow the use case, without the need for runtime checks.

To me, that sounds like your "There is no point in allowing `R.__` as a last argument"-example — if thats the intention of the library: Just don't add an overload for it.

---

What makes Placeholders so tricky to type is:  You need to either
**a)** type each use case by hand
**b)** have some quite complex generic typing to automatically generate the overloads.

**a)** will inevitably miss some use case and will still be quite a lot of work. It will also be super hard to maintain.

**b)** seems to be the easier approach, but it will sooner or later reach its limitation — at the latest when generics come into play (probably making you have to fallback on a)). It most likely will also make the resulting types super hard to read — probably for everybody, even people who never want to use Placeholders.

---

My assessment would be that adding full Placeholder support would most definitely be fantastic for people who actually use Placeholders **and** TypeScript — if we ever get it to fully work and not fail halfway through. But it will be an immense amount of work on the **whole** library (just for one feature) and make maintaining much harder in the future.

For me this brings us back to the question: is it really worth it? All the work put into supporting Placeholders, will not only not at all benefit people who don't use. It will probably make the library more error prone and less well maintained and up to date, then if we just dropped support.

---

I agree with @rbuckton, that we should look at the ramifications, of dropping support for Placeholder. But we should also look at the ramifications of continuing support for it or even expanding it. Right now it looks like, Placeholder is of little use, when used with TypeScript due to how little support there is as of now.

Maybe someone could show some examples of how function types would change with full Placeholder support?

**Comment by CrossEye** [comment]
> Maybe someone could show some examples of how function types would change with full Placeholder support?

I can't do that; I don't have the necessary TS skills.  But it will multiply the number of signatures greatly, even if  Ramda disallows a final placeholder argument.   If it doesn't, I'm not sure what can be done, as this is perfectly legal in Ramda:

```js
foo (arg1) (__) (__) (__) (__) (__) (__) (__) (__) (__) (__) (__) (__) (__) (__) (__) (__) (__) (__) (arg2)
//         `--------------------------------------- n times ----------------------------------------'
```

for any non-negative integer, `n`.

So if `foo` is a Ramda-curried version of `(a, b, c) -> d`, it supports any of these calls:

```js
foo (a, b, c)
foo (a, b) (c)
foo (a) (b, c)
foo (a) (b) (c)
foo (__, b) (a, c)
foo (__, b) (a) (c)
foo (__, b) (__, c) (a)
foo (__, b, c) (a)
foo (__, __, c) (a, b)
foo (__, __, c) (a) (b)
foo (__, __, c) (__, b) (a)
foo (a, __, c) (b)
```

(and I'm likely missing some.)

This does not include the stupid calls such as `f (a) () () (b) () () () () (c)`, which I think we can safely ignore.  Anyone doing that deserves to get a hand slapped.

Am I correct that even if a signature is not specified in the typings file, the user can overwrite the type of any given expression as needed?  So if the Ramda-supported type was not supplied by TS, the user could still describe the type manually?

**Comment by essenmitsosse** [comment]
I actually played around a little bit, and it seems with recursive types implementing Placeholder would actually be possible — **including** all the weird edge cases.

[Playground](https://www.typescriptlang.org/play?jsx=0#code/C4TwDgpgBA+jUF4oG9SQFxQOQAUA2AhgMYQAWA9ngCYQBOWAvgNwBQR5AdgM7CwyZxEKZixYBLDsDoAzYtABi5cihZQ1UABQBKTIvKr1G-nx1Q9AGgNqjAmOb63TFq1AA8AFQB8NvvYBGmABKEARUnHggAIK0tAQgHp5OSjAJLgkaBJjuSeQAjKnqbl4ZWf5BIWEcEdGx8V6mANruALqsDKISUrSyJGZK+V4qhdq6-QWGxnA5A56WwwFQwaHhUTFxCY0tbR2SMnJ95CmDyC4jB0eep5MwORdzhplQ2ZhNzSztbJw8UNJKo8pIbSITzCVifbi8X5CX7kDTIGHoXIMLTg760CBcACueF4SF+Gga8L+UAATAxmijUZDDtClEY7FAGlgYVgKVSoOisTiSbSYBoAES5fmU9gQjkY7HAADMtNhU209K0ivp9gawFomIgFI0sjwXAgWiAA)

Downsides are:

- Types have to be written by hand (at least partly, there might be some abstractions) — which would lead to an explosion of complexity
- The resulting types **work** very well. Actually exactly as expected. **But** they are pretty unreadable to the user, since they are obfuscated behind an interface (I don't know how to make the overloads visible, while still making them recursive)

**Comment by essenmitsosse** [comment]
If we wouldn't have to worry about generics you could type it like that, which would result in quite decent types:

(Playground)[https://www.typescriptlang.org/play?jsx=0#code/C4TwDgpgBA+jUF4oG9SQFxQOQAUA2AhgMYQAWA9ngCYQBOWAvgNwBQR5AdgM7CwyZxEKZixYAzAK4ciwAJacoY8uQAUASkxoI5MYuWtJ0uQqWr+fDVC0695A1JnyOtlebgAaPgJiXru0-ZGTrYAPAAqAHyu3p4ARpgAShAEVJx4IACCtLQEICE8tLIcAOYRvuDa-sow4g7GzqbhUQSYYeWQNqYAjLVBJspNKi1QYXGJyakc6Vk5eQVFpZYA2mEAuqKGjv3kXeqaFZ3KPZv1trtuPvsdVTu9Ww1HKvFQSSlpmdm5+cCFJWWYK1WUAA9MCoCEALRQiFQADq0CIBA4WF4xBIXC4IygpDo0DwsgA1tBgKQCMANnVgqYYHsrAcbjUTlTqtELFdKrZGZTtjBBsM2gC1qI2JweHohKYVMhTOgugw1CLuLxaBAuBI8LwkEoVEssKZYgR6KsFYqxUp4FrlK4YJ5daYsMbTcrVergAAmCXkGkAIi63pN7CVUBVao1AGZPWYfOprWpY9bbXqJAAvWIph1xsQEPBcCBqIA]

But there are barely Ramda functions that wouldn't need generics. :/

**Comment by CrossEye** [comment]
> Worth noting that while it is *legal* - is it a case that Ramda wants to support?

We could change this behavior, while it is intentional, it is mostly just there for logical compatibility and not for practical work.  I would also have no objection to having TS reject it even if underneath Ramda still allows it.

The reasoning is simple enough for empty argument lists returning a function equivalent to the one that was called.  If we start with, say, four formal parameters, and we supply four of them, we're done and we call the function.  If we supply three of them, we get back a function expecting the remaining one.  If we supply two, we get back a function expecting the remaining two.  If we supply one, we get back a function expecting the remaining three.  So it seems eminently logical that if we supply zero, we get back a function expecting all four.

For placeholder, it's a very similar case.  Again, with four formal parameters, if we supply non-placeholder values at indices `0` and `3`, we get back a function that will fill indices `1` and `2`.  If we supply non-placeholder values at index 2, we get back a function that will fill indices `0`, `1`, and `3`.  If we supply non-placeholder values at index `3`, we get back a function that will fill indices `0`, `1`, and `2`.  So if we supply no non-placeholder values, we get back a function that will fill indices `0`, `1`, `2`, and `3`.

This feels much like the fact that `[] .every (anyPredicate) //=> true` or `sum ([]) //=> 0`.  It's logically consistent and occasionally useful.

The big problem for Ramda if we were to change this is that our only alternative here would likely be to throw an exception, and Ramda, whose main focus is on composability, is very loathe to throw.  But we can work that out.  I wouldn't bother, though, if this is something that TS can handle without crazy gymnastics.

**Comment by somebody1234** [comment]
TL;DR - sorry, bad wording - I meant more along the lines of "is this an intended usecase of Ramda" rather than "maybe Ramda should go out of its way to disallow it"

IMO it's less about "the types should reflect the semantics of Ramda 1:1" and more "the types should reflect Ramda's intended usecases" -
In situations like these it does feel like doing certain things `R.foo()()()()`, `R.foo(__)(__)(__)(__)` etc. are probably almost always a mistake

Or in other words, the way I see it, TypeScript's intention is not to be a type system so much as it is meant to be a tool to help in writing correct code

**Comment by somebody1234** [comment]
Note though, I think to support placeholder well, we'd need to add a huge number of new interfaces (for all possible combinations of  placeholders and filled-in parameters) - I'm not sure if it's possible to make a helper type but I wouldn't be optimistic because said helper type would have to work correctly with generics

So the question is, is the relatively low fraction (not sure how to estimate actually) of users using placeholder, worth making the types several times bigger (and potentially less ergonomic)

**Comment by CrossEye** [comment]
Is there some well-respected introduction to TS typings files?

I think I'm still missing a lot of the big picture.  I'm not sure how much time and energy I want to invest in learning about them, but if I do choose to put some real effort into it, what's a good place to start?  I've seen plenty of the *TS for JS developers* articles and the *I never thought I'd switch, but now I'm so happy ones*, and none have really struck any chords for me.  If I'm going to dig in, it's probably from this angle.  *How can we handle JS's trickier signatures?*

**Comment by somebody1234** [comment]
I think [the handbook](https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html) is what's generally recommended - however note that the other sections (on TypeScript in general) are probably useful too - for the most part this is TypeScript definitions, anything specific to typings files that may have needed to be done has already been done.

Supposedly it takes 30 minutes to go through, although I suspect it'd take quite a bit longer to absorb - and even longer still to play around with specific typings(/constructs) to get a feel of the strengths and weaknesses of each
